man/hashdeep.1
.TH HASHDEEP "1" "v4.4 \- 29 Jan 2014" "AFOSI" "United States Air Force"

.SH NAME
hashdeep \- Compute, compare, or audit multiple message digests

.SH SYNOPSIS
.B hashdeep 
-V | -h
.br
.B hashdeep
[-c <alg1>[,<alg2>]] [-k <file>] [-i <size>] [\-f <file>] 
[\-o <fbcplsde>] [-amxwMXreEspblvv] [\-F<bum>] [\-j <num>] [\fBFILES\fR]


.SH DESCRIPTION
.PP
Computes multiple hashes, or message digests, 
for any number of files while 
optionally recursively digging through the directory structure.
By default the program computes MD5 and SHA-256 hashes, equivalent
to \-c md5,sha256.
Can also take a list of known hashes and display the filenames
of input files whose hashes either do or do not match any of the
known hashes.
Can also use a list of known hashes to audit a set of FILES.
Errors are reported to standard error. If no FILES are specified,
reads from standard input.



.TP
\fB\-c <alg1>[,<alg2>...]\fR
Computation mode. Compute hashes of FILES using the algorithms 
specified. Legal values are md5, sha1, sha256, tiger, and
whirlpool. 


.TP
\fB\-k \fR
Load a file of known hashes. 
This flag is required when using any of the matching or audit
modes (i.e. \-m, \-x, \-M, \-X, or \-a)
This flag may be used more than once
to add multiple sets of known hashes. 

Loading sets with different hash algorithms
can sometimes generate spurrious hash collisions. For
example, let's say we have two hash sets, A and B, 
which have some overlapping files. For example, the file
/usr/bin/bad is in both sets. In A we've recorded the MD5 and SHA-256.
In B we've recorded the MD5, SHA-1, and SHA-256. Because these
two records are different, they will both be loaded. When the 
program computes all three hashes and compares them to the set 
of knowns, we will get an exact match from the record in B 
and a collision from the record in A. 


.TP
\fB\-a \fR
Audit mode. Each input file is compared against the set of knowns. 
An audit is said to pass if each input file is matched against exactly
one file in set of knowns. Any collisions, new files, or missing
files will make the audit fail. Using this flag alone produces a
message, either "Audit passed" or "Audit Failed". Use the
verbose modes, \-v, for more details. Using \-v prints the number 
of files in each category. Using \-v a second time prints any
discrepancies. Using \-v a third time prints the results for every file
examined and every known file.
.br
Due to limitations in the program, any filenames with Unicode characters
will appear to have moved during an audit. See the section "UNICODE
SUPPORT" below.

.TP
\fB\-m \fR
Positive matching, requires at least one use of the \-k flag.
The input files are examined one at a time, and only those files that match
the list of known hashes are output. The only acceptable format
for known hashes is the output of previous hashdeep runs.
.br
\fB\fR
If standard input is used with the -m flag, displays "stdin"
if the input matches one of the hashes in the list of known hashes. If the
hash does not match, the program displays no output.
.br
\fB\fR
This flag may not be used in conjunction with the \-x, \-X, or \-a flags.
See the section "UNICODE SUPPORT" below.

.TP
\fB\-x \fR
Negative matching.
Same as the \-m flag above, but does negative matching. That is, only 
those files NOT in the list of known hashes are displayed. 
.br
\fB\fR
This flag may not be used in conjunction with the \-m, \-M, or \-a flags.
See the section "UNICODE SUPPORT" below.

.TP
\fB\-f\fR <file>
Takes a list of files to be hashed from the specified file. Each
line is assumed to be a filename. This flag can only be used once
per invocation. If it's used a second time, the second instance will
clobber the first. 
.br
Note that you can still use other flags, such as the \-m or \-x modes,
and submit additional FILES on the command line.

.TP
\fB\-w \fR
When used with positive matching modes (\-m,\-M) displays the filename
of the known hash that matched the input file. 
See the section "UNICODE SUPPORT" below.

.TP
\fB\-M\fR and \fB-X\fR
Same as \-m and \-x above, but displays the hash for each file that 
does (or does not) match the list of known hashes. 


.TP
\fB\-r\fR
Enables recursive mode. All subdirectories are traversed. Please note
that recursive mode cannot be used to examine all files of a given 
file extension. For example, calling hashdeep -r *.txt will examine
all files in \fIdirectories\fR that end in .txt. 


.TP
\fB\-e\fR
Displays a progress indicator and estimate of time
remaining for each file being processed. Time estimates for files
larger than 4GB are not available on Windows. This mode may not be
used with th \-p mode.

.TP
\fB\-E\fR
When in audit mode, performs case insensitive matching of filenames.
For example, \\foo\\bar will match to \\Foo\\BAR. This can be important
on Windows systems, where filenames are case insensitive.

.TP
\fB\-i <size> \fR
Size threshold mode. Only hash files smaller than the given the 
threshold. Sizes may be specified 
using IEC multipliers b,k,m,g,t,p, and e.


.TP
\fB\-o\fR <bcpflsd>
Enables expert mode. Allows the user specify which (and only which) types of
files are processed. Directory processing is still controlled with the
\-r flag. The expert mode options allowed are:
.br
f \- Regular files
.br
b \- Block Devices
.br
c \- Character Devices
.br
p \- Named Pipes
.br
l \- Symbolic Links
.br
s \- Sockets
.br
d \- Solaris Doors
.br
e \- Windows PE executables


.TP
\fB\-s\fR
Enables silent mode. All error messages are supressed.


.TP
\fB\-p\fR
Piecewise mode. Breaks files into chunks before hashing. Chunks
may be specified using IEC multipliers b,k,m,g,t,p, and e. (Never let
it be said that the author didn’t plan ahead.) 


.TP
\fB\-b\fR
Enables bare mode. Strips any leading directory information from 
displayed filenames.
This flag may not be used in conjunction with the \-l flag.

.TP
\fB\-l\fR
Enables relative file paths. Instead of printing the absolute path for
each file, displays the relative file path as indicated on the command 
line. This flag may not be used in conjunction with the \-b flag.

.TP
\fB\-v\fR
Enables verbose mode. Use again to make the program more verbose. 
This mostly changes the behavior of the audit mode, \-a.

.TP
\fB-jnn\fR
Controls multi-threading. By default the program will create one
producer thread to scan the file system and one hashing thread per CPU
core. Multi-threading causes output filenames to be in
non-deterministic order, as files that take longer to hash will be
delayed while they are hashed. If a deterministic order is required,
specify \fB-j0\fR to disable multi-threading

.TP
\fB-d\fR
Output in Digital Forensics XML (DFXML) format.

.TP
\fB-u\fR
Quote Unicode output. For example, the snowman is shown as
\fBU+C426\fR.

.TP
\fB-F<bum>\fR
Specifies the input mode that is used to read files. The default is
\fB-Fb\fR (buffered I/O) which reads files with fopen(). Specifying
\fB-Fu\fR will use unbuffered I/O and read the file with
open(). Specifying \fB-Fm\fR will use memory-mapped I/O which will be
faster on some platforms, but which (currently) will not work with
files that produce I/O errors.

.TP
\fB-R\fR
Don't follow symlinks, instead hash the output of readlink. (Not available on
Windows)



.TP
\fB\-h\fR
Show a help screen and exit.

.TP
\fB\-V\fR
Show the version number and exit.



.SH UNICODE SUPPORT
As of version 3.0 the program supports Unicode characters in filenames
on Microsoft Windows systems for filenames specified on the command
line with globbing (e.g. *), for files specified with the
\fB-f\fR of files to hash, and for files read from directories using
the \fB-r\fR option.

By default all program input and output
should be in UTF-8.  The program automatically converts this to UTF-16
for opening files). 

On Unix/Linux/MacOS, you should use a terminal emulator that supports
UTF-8 and UTF-8 characters in filenames will be properly displayed.

On Windows, the programs do not display Unicode characters on the console.
You must either redirect output to a file and open the
file with Wordpad (which can display Unicode), or you must specify the
\fB-u\fR option to quote Unicode using standard \fBU+XXXX\fR notation.

Currently the file name of a file containing known hashes may not be
specified as a Unicode filename, but you can specify the name using
tab completion or an asterisk (e.g. md5deep -m *.txt where there is
only one file with a .txt extension).

.SH RETURN VALUE
Returns a bit-wise value based on the success of the operation and the
status of any matching operations.
.PP
.TP
0
Success. Note that the program considers itself successful even when it
encounters read errors, permission denied errors, or finds directories
when not in recursive mode.
.TP
1
Unused hashes. Under any of the matching modes, returns this 
value if one or more of the
known hashes was not matched by any of the input files.
.TP
2
Unmatched inputs. Under any of the matching modes, returns this value
if one or more of the input values did not match any of the known hashes. 
.TP
64
User error, such as trying to do both positive and negative matching at 
the same time. 
.TP
128
Internal error, such as memory corruption or uncaught cycle.
All internal errors should
be reported to the developer! See the section "Reporting Bugs" below.


.SH AUTHOR
hashdeep was written by Jesse Kornblum, research@jessekornblum.com,
and Simson Garfinkel.

.SH KNOWN ISSUES
Using the \-r flag cannot be used to recursively process all files 
of a given extension in a directory. This is a feature, not a bug. 
If you need to do this, use the \fBfind\fR(1) command.

The program will fail if you attempt to compare 2^64 or more input
files against a set of known files. 


.SH REPORTING BUGS
We take all bug reports \fIvery\fR seriously. Any bug that jeopardizes the
forensic integrity of this program could have serious consequences on 
people's lives. When submitting a bug report, please include a description
of the problem, how you found it, and your contact information.
.PP
Send bug reports to the author at the address above.

.PP
.SH COPYRIGHT
This program is a work of the US Government. In accordance with 17 USC 105,
copyright protection is not available for any work of the US Government.
This program is PUBLIC DOMAIN. Portions of this program contain code
that is licensed under the terms of the General Public License (GPL).
Those portions retain their original copyright and license. See the file
COPYING for more details.
.PP
There is NO warranty for this program; 
not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

.SH SEE ALSO
More information and installation instructions can be found in the README 
file. Current versions of both documents can be found on the project homepage: 
http://md5deep.sourceforge.net/
.PP
The MD5 specification, RFC 1321, is available at
.br
http://www.ietf.org/rfc/rfc1321.txt
.PP
The SHA-1 specification, RFC 3174, is available at
.br
http://www.faqs.org/rfcs/rfc3174.html
.PP
The SHA-256 specification, FIPS 180-2, is available at
.br
http://csrc.nist.gov/publications/fips/fips180-2/fips180-2.pdf
.PP
The Tiger specification is available at
.br
http://www.cs.technion.ac.il/~biham/Reports/Tiger/
.PP
The Whirlpool specification is available at
.br
http://planeta.terra.com.br/informatica/paulobarreto/WhirlpoolPage.html


NEWS
** Changes in version 4.4 (29 Jan 2014)

* New Features

      Added case insensitive auditing with -E mode

* Bug Fixes

      Added checks against user errors on Hashdeep command line

      Fixed bug for known filenames with the directory in the filename

      Fixed bug for double dots in the filename



** Changes in version 4.3 (24 Oct 2012)

* Bug Fixes

      Fixed check for endian-ness, affecting hash generation on
      big-endian platforms.

      Fixed minor bugs related to OpenSolaris.



** Changes in version 4.2 (11 Jun 2012)

* Bug Fixes

      Fixed padding in Tiger hashes for large files



** Changes in version 4.1.1 (5 Jun 2012)

* Bug Fixes

      Fixed hashdeep bug for known files with commas in their names.
      Cleaned up usage message


** Changes in version 4.1.0 (14 Feb 2012)

* New Features

      Added expert mode option to parse Windows PE files

* Bug Fixes

      Fixed junction point handling on Win32



** Changes in version 4.0.1 (22 Jan 2012)

* Bug Fixes

        Fixed hang on DFXML generation on Win32
	Fixed incorrect hashes via stdin on Win32
	Fixed "Too many open files" error on OS X
	Doc files in Win32 have been corrected.



** Changes in version 4.0 (6 Jan 2012)

* New Features

      	Rewrote most of the program in C++.

	Enabled multiprocessor support on all platforms.

	Removed ten character limit on file size mode

* Bug Fixes

	Too numerous to count!



** Changes in version 3.9.1 (31 May 2011)

* Bug Fixes

      Fixed a cosmetic bugs regarding processing standard input
      in piecewise mode.

      Fixed a bug regarding processing known hash files which have
      parenthesis in the filename.

      Clarified documentation regarding multipliers in piecewise
      mode. They are IEC multipliers, or blocks of 1,024. Side note:
      also spelled 'multipliers' correctly.



** Changes in version 3.9 (21 Apr 2011)

* Bug Fixes

      Fixed a critical bug in hashdeep regarding processing
      standard input



** Changes in version 3.8 (15 Apr 2011)

* New Features

      Added -f mode to get the list of input files from
      a file on the disk.

      Added 64-bit executables for Windows.


* Bug Fixes

      Fixed a cosmetic bug regarding the error messages for
      piecewise mode.




** Changes in version 3.7 (18 Dec 2010)

* New Features

      Added triage mode, -Z flag, to compute file size and partial
      hashes in addition to full file hash.

      Updated Windows cross-compiler to MinGW-4.2.1


* Bug Fixes

      Fixed hashdeep piecewise mode to include the correct size
      of each piece being hashed. This enables audit mode to
      work correctly.

      Addressed display problem for time estimates of files with
      long filenames in the current directory.

      Added "Input/Output error" to the list of fatal errors when
      processing files




** Changes in version 3.6 (23 Mar 2010)

* New Features

      Added expert mode found in md5deep to hashdeep.

* Bug Fixes

      Fixed error checking in helper routines.




** Changes in version 3.5.1 (30 Dec 2009)

* Bug Fixes

      Fixed processing of zero byte files.




** Changes in version 3.5 (22 Dec 2009)

* New Features

      Added mode to load list of files to process from a file (-f mode)

* Bug Fixes

      Fixed file size computation for piecewise hashing of standard input.

      Fixed error when auditing a directory to check for identical copies
      of a file if one copy is removed.



** Changes in version 3.4 (11 Jun 2009)

* Bug Fixes

      Fixed directory handle leak during recursive parsing on Windows

      Fixed -n mode (display unused hashes)



** Changes in version 3.3 (4 Apr 2009)

* New Features

* Bug Fixes

      Skip all reparse points (symbolic links and junction points)
      on Windows. Required for sane parsing of Windows Vista or 7 systems.

      Fixed handling of 64-bit timestamps on Windows.




** Changes in version 3.2 (25 Jan 2009)

* New Features

	Added Comma Separated Value (CSV) mode.


* Bug Fixes

	Fixed support for 64-bit files on Windows

	Fixed hashdeep input file parsing to allow filenames with
	commas and properly handle other kinds of errors. Also
	modified the invocation listing in output files such that
        even long lines can be successfully parsed.




** Changes in version 3.1 (23 Jul 2008)

* New Features

	Added size threshold modes to only process files smaller
	than a given size.

	Added a timestamp mode that records the creation time time
	for each file on Win32, the change time on all other operating
	systems.

	Streamlined file size computation process, which makes
	the programs about 15% faster.


* Bug Fixes

	Corrected time estimates for large files (e.g. files which
        require more than one day).




** Changes in version 3.0 (10 May 2008)

* New Features

	Added hashdeep program to support multihashing and
        hash file auditing

	Added support for new iLook style hashes

* Bug Fixes

	Fixed obscure bug that caused a crash (double free) when
	attempting to check a very small file for EnCase hashes



** Changes in version 2.0.1 (31 Oct 2007)

* Bug Fixes

	Added support for older Linux systems that don't have the
	latest and greatest ioctl functions.



** Changes in version 2.0 (15 Oct 2007)

* New Features

	Using GNU Autotools for configuration and compilation. This
	should help avoid platform specific issues such as SHA-1
	problems previously found on 64-bit versions of AIX.

	Added support for files with Unicode characters in their
	filenames on Microsoft Windows.

	Added support for EnCase hash sets (.hash files).

	Updated web site and quick start guide

	Slightly reduced the size of all of the executables by removing
	duplicated code.

* Bug Fixes

	Fixed time estimation mode for block devices on OS X and Linux

	Fixed cosmetic error where estimated time remaining mode is
	being used in conjunction with piecewise hashing. Time estimates

hashdeep/hashdeep.cpp
// HASHDEEP
// $Id$

extern "C" {

#ifdef MD5DEEP
#include "md5deep.h"
#else
#include "main.h"
#endif

};

#ifdef _WIN32 
// This can't go in main.h or we get multiple definitions of it
// Allows us to open standard input in binary mode by default 
// See http://gnuwin32.sourceforge.net/compile.html for more 
int _CRT_fmode = _O_BINARY;
#endif


// So that the usage message fits in a standard DOS window, this
// function should produce no more than 22 lines of text.
static void usage(state *s)
{
  print_status("%s version %s by %s.",__progname,VERSION,AUTHOR);
  print_status("%s %s [-c <alg>] [-k <file>] [-amxwMXrespblvv] [-V|-h] [-o <mode>] [FILES]",CMD_PROMPT,__progname);

  print_status("");

  print_status("-c <alg1,[alg2]> - Compute hashes only. Defaults are MD5 and SHA-256");
  fprintf(stdout,"     legal values are ");
  for (int i = 0 ; i < NUM_ALGORITHMS ; i++)
    fprintf(stdout,"%s%s",s->hashes[i]->name,(i+1<NUM_ALGORITHMS)?",":NEWLINE);

  print_status("-a - audit mode. Validates FILES against known hashes. Requires -k");
  print_status("-d - output in DFXML (Digital Forensics XML)");
  print_status("-m - matching mode. Requires -k");
  print_status("-x - negative matching mode. Requires -k");
  print_status("-w - in -m mode, displays which known file was matched");
  print_status("-M and -X act like -m and -x, but display hashes of matching files");
  print_status("-k - add a file of known hashes");
  print_status("-r - recursive mode. All subdirectories are traversed");
  print_status("-e - compute estimated time remaining for each file");
  print_status("-s - silent mode. Suppress all error messages");
  print_status("-p - piecewise mode. Files are broken into blocks for hashing");
  print_status("-b - prints only the bare name of files; all path information is omitted");
  print_status("-l - print relative paths for filenames");
  print_status("-i - only process files smaller than the given threshold");
  print_status("-o - only process certain types of files. See README/manpage");
  print_status("-v - verbose mode. Use again to be more verbose.");
  print_status("-V - display version number and exit");
}


static void check_flags_okay(state *s)
{
  sanity_check(s,
	       (((s->primary_function & primary_match) ||
		 (s->primary_function & primary_match_neg) ||
		 (s->primary_function & primary_audit)) &&
		!s->hashes_loaded),
	       "Unable to load any matching files");

  sanity_check(s,
	       (s->mode & mode_relative) && (s->mode & mode_barename),
	       "Relative paths and bare filenames are mutally exclusive");
  
  /* Additional sanity checks will go here as needed... */
}



static int 
add_algorithm(state *s, 
	      hashname_t pos,
	      const char *name, 
	      uint16_t len, 
	      int ( *func_init)(void *),
	      int ( *func_update)(void *, unsigned char *, uint64_t ),
	      int ( *func_finalize)(void *, unsigned char *),
	      int inuse)
{
  if (NULL == s)
    return TRUE;

  s->hashes[pos] = (algorithm_t *)malloc(sizeof(algorithm_t));
  if (NULL == s->hashes[pos])
    return TRUE;

  s->hashes[pos]->name = strdup(name);
  if (NULL == s->hashes[pos]->name)
    return TRUE;

  s->hashes[pos]->known = (hashtable_t *)malloc(sizeof(hashtable_t));
  if (NULL == s->hashes[pos]->known)
    return TRUE;

  hashtable_init(s->hashes[pos]->known); 

  s->hashes[pos]->hash_sum = (unsigned char *)malloc(len * 2);
  if (NULL == s->hashes[pos]->hash_sum)
    return TRUE;

  s->hashes[pos]->hash_context = malloc(ALGORITHM_CONTEXT_SIZE);
  if (NULL == s->hashes[pos]->hash_context)
    return TRUE;
  
  s->hashes[pos]->f_init      = func_init;
  s->hashes[pos]->f_update    = func_update;
  s->hashes[pos]->f_finalize  = func_finalize;
  s->hashes[pos]->byte_length = len;
  s->hashes[pos]->inuse       = inuse;

  return FALSE;
}


int setup_hashing_algorithms(state *s)
{
  if (NULL == s)
    return TRUE;

  /* The DEFAULT_ENABLE variables are in main.h */

  if (add_algorithm(s,
		    alg_md5,
		    "md5",
		    32,
		    hash_init_md5,
		    hash_update_md5,
		    hash_final_md5,
		    DEFAULT_ENABLE_MD5))
    return TRUE;
  if (add_algorithm(s,
		    alg_sha1,
		    "sha1",
		    40,
		    hash_init_sha1,
		    hash_update_sha1,
		    hash_final_sha1,
		    DEFAULT_ENABLE_SHA1))
    return TRUE;
  if (add_algorithm(s,
		    alg_sha256,
		    "sha256",
		    64,
		    hash_init_sha256,
		    hash_update_sha256,
		    hash_final_sha256,
		    DEFAULT_ENABLE_SHA256))
    return TRUE;
  if (add_algorithm(s,
		    alg_tiger,
		    "tiger",
		    48,
		    hash_init_tiger,
		    hash_update_tiger,
		    hash_final_tiger,
		    DEFAULT_ENABLE_TIGER))
    return TRUE;
  if (add_algorithm(s,
		    alg_whirlpool,
		    "whirlpool",
		    128,
		    hash_init_whirlpool,
		    hash_update_whirlpool, 
		    hash_final_whirlpool,
		    DEFAULT_ENABLE_WHIRLPOOL))
    return TRUE;

  return FALSE;
}


void clear_algorithms_inuse(state *s)
{
    //hashname_t i;

  if (NULL == s)
    return;

  for (int i = 0 ; i < NUM_ALGORITHMS ; ++i)
  { 
    if (s->hashes[i] != NULL)
      s->hashes[i]->inuse = FALSE;
  }
}


static int initialize_hashing_algorithms(state *s)
{
    //hashname_t i;
  
  for (int i = 0 ; i < NUM_ALGORITHMS ; ++i)
  {
    if (NULL == s->current_file->hash[i])
    {
      s->current_file->hash[i] = (char *)malloc(sizeof(char) * s->hashes[i]->byte_length * 2);
      if (NULL == s->current_file->hash[i])
	fatal_error(s,"%s: Out of memory", __progname);
    }
  }

  return FALSE;
}


static int parse_hashing_algorithms(state *s, char *val)
{
  uint8_t i;
  char **ap, *buf[MAX_KNOWN_COLUMNS];

  if (NULL == s || NULL == val)
    return TRUE;

  for (ap = buf ; (*ap = strsep(&val,",")) != NULL ; )
    if (*ap != '\0')
      if (++ap >= &buf[MAX_KNOWN_COLUMNS])
	break;

  for (i = 0 ; i < MAX_KNOWN_COLUMNS && buf[i] != NULL ; i++)
  {
    if (STRINGS_CASE_EQUAL(buf[i],"md5"))
      s->hashes[alg_md5]->inuse = TRUE;
    
    else if (STRINGS_CASE_EQUAL(buf[i],"sha1") || 
	     STRINGS_CASE_EQUAL(buf[i],"sha-1"))
      s->hashes[alg_sha1]->inuse = TRUE;
    
    else if (STRINGS_CASE_EQUAL(buf[i],"sha256") || 
	     STRINGS_CASE_EQUAL(buf[i],"sha-256"))
      s->hashes[alg_sha256]->inuse = TRUE;
    
    else if (STRINGS_CASE_EQUAL(buf[i],"tiger"))
      s->hashes[alg_tiger]->inuse = TRUE;
    
    else if (STRINGS_CASE_EQUAL(buf[i],"whirlpool"))
      s->hashes[alg_whirlpool]->inuse = TRUE;
    
    else if (STRINGS_CASE_EQUAL(buf[i],"all"))
    {
	//hashname_t count;
      for (int count = 0 ; count < NUM_ALGORITHMS ; ++count)
	s->hashes[count]->inuse = TRUE;
      return FALSE;
    }
      
    else
    {
      print_error(s,"%s: Unknown algorithm: %s", __progname, buf[i]);
      try_msg();
      exit(EXIT_FAILURE);
    }    
  }
    
  return FALSE;
}


static int process_command_line(state *s, int argc, char **argv)
{
  int i;
  
  while ((i=getopt(argc,argv,"do:I:i:c:MmXxtablk:resp:wvVh")) != -1)
  {
    switch (i)
    {
    case 'o':
      s->mode |= mode_expert; 
      setup_expert_mode(s,optarg);
      break;

    case 'I': 
      s->mode |= mode_size_all;
      // Note no break here;
    case 'i':
      s->mode |= mode_size;
      s->size_threshold = find_block_size(s,optarg);
      if (0 == s->size_threshold)
      {
	print_error(s,"%s: Requested size threshold implies not hashing anything",
		    __progname);
	exit(STATUS_USER_ERROR);
      }
      break;

    case 'c': 
      s->primary_function = primary_compute;
      /* Before we parse which algorithms we're using now, we have 
	 to erase the default (or previously entered) values */
      clear_algorithms_inuse(s);
      if (parse_hashing_algorithms(s,optarg))
	fatal_error(s,"%s: Unable to parse hashing algorithms",__progname);
      break;
      
    case 'd': s->dfxml = 1; break;
    case 'M': s->mode |= mode_display_hash;	  
    case 'm': s->primary_function = primary_match;      break;
      
    case 'X': s->mode |= mode_display_hash;
    case 'x': s->primary_function = primary_match_neg;  break;
      
    case 'a': s->primary_function = primary_audit;      break;
      
      // TODO: Add -t mode to hashdeep
      //    case 't': s->mode |= mode_timestamp;    break;

    case 'b': s->mode |= mode_barename;     break;
    case 'l': s->mode |= mode_relative;     break;
    case 'e': s->mode |= mode_estimate;     break;
    case 'r': s->mode |= mode_recursive;    break;
    case 's': s->mode |= mode_silent;       break;
      
    case 'p':
      s->mode |= mode_piecewise;
      s->piecewise_size = find_block_size(s, optarg);
      if (0 == s->piecewise_size)
	fatal_error(s,"%s: Piecewise blocks of zero bytes are impossible", 
		    __progname);
      
      break;
      
    case 'w': s->mode |= mode_which;        break;
      
    case 'k':
      switch (load_match_file(s,optarg))
	{
	case status_ok: 
	  s->hashes_loaded = TRUE;
	  break;
	  
	case status_contains_no_hashes:
	  /* Trying to load an empty file is fine, but we shouldn't
	     change s->hashes_loaded */
	  break;
	  
	case status_contains_bad_hashes:
	  s->hashes_loaded = TRUE;
	  print_error(s,"%s: %s: contains some bad hashes, using anyway", 
		      __progname, optarg);
	  break;
	  
	case status_unknown_filetype:
	case status_file_error:
	  /* The loading code has already printed an error */
	    break;
	    
	default:
	  print_error(s,"%s: %s: unknown error, skipping", __progname, optarg);
	  break;
	}
      break;
      
    case 'v':
      if (s->mode & mode_insanely_verbose)
	print_error(s,"%s: User request for insane verbosity denied", __progname);
      else if (s->mode & mode_more_verbose)
	s->mode |= mode_insanely_verbose;
      else if (s->mode & mode_verbose)
	s->mode |= mode_more_verbose;
      else
	s->mode |= mode_verbose;
      break;
      
    case 'V':
      print_status("%s", VERSION);
      exit(EXIT_SUCCESS);
	  
    case 'h':
      usage(s);
      exit(EXIT_SUCCESS);
      
    default:
      try_msg();
      exit(EXIT_FAILURE);
    }            
  }
  
  check_flags_okay(s);

  return FALSE;
}


static int initialize_state(state *s) 
{
  if (NULL == s)
    return TRUE;

  if (setup_hashing_algorithms(s))
    return TRUE;

  MD5DEEP_ALLOC(file_data_t,s->current_file,1);
  MD5DEEP_ALLOC(TCHAR,s->full_name,PATH_MAX);
  MD5DEEP_ALLOC(TCHAR,s->short_name,PATH_MAX);
  MD5DEEP_ALLOC(TCHAR,s->msg,PATH_MAX);
  MD5DEEP_ALLOC(unsigned char,s->buffer,MD5DEEP_IDEAL_BLOCK_SIZE);

  s->known            = NULL;
  s->last             = NULL;
  s->piecewise_size   = 0;
  s->hash_round       = 0;
  s->primary_function = primary_compute;
  s->mode             = mode_none;
  s->hashes_loaded    = FALSE;
  s->banner_displayed = FALSE;
  s->block_size       = MD5DEEP_IDEAL_BLOCK_SIZE;
  s->size_threshold   = 0;
  s->expected_columns = 0;

  return FALSE;
}


#ifdef _WIN32
static int prepare_windows_command_line(state *s)
{
  int argc;
  TCHAR **argv;

  argv = CommandLineToArgvW(GetCommandLineW(),&argc);
  
  s->argc = argc;
  s->argv = argv;

  return FALSE;
}
#endif


int main(int argc, char **argv)
{
  int count, status = EXIT_SUCCESS;
  TCHAR *fn;
  state *s;

  /* Because the main() function can handle wchar_t arguments on Win32,
     we need a way to reference those values. Thus we make a duplciate
     of the argc and argv values. */ 

#ifndef __GLIBC__
  __progname  = basename(argv[0]);
#endif

  s = (state *)malloc(sizeof(state));
  if (NULL == s)
  {
    // We can't use fatal_error because it requires a valid state
    print_status("%s: Unable to allocate state variable", __progname);
    return EXIT_FAILURE;
  }

  if (initialize_state(s))
  {
    print_status("%s: Unable to initialize state variable", __progname);
    return EXIT_FAILURE;
  }

  process_command_line(s,argc,argv);
 
  if (initialize_hashing_algorithms(s))
    return EXIT_FAILURE;
   
  if (primary_audit == s->primary_function)
    setup_audit(s);

#ifdef _WIN32
  if (prepare_windows_command_line(s))
    fatal_error(s,"%s: Unable to process command line arguments", __progname);
#else
  s->argc = argc;
  s->argv = argv;
#endif

  MD5DEEP_ALLOC(TCHAR,s->cwd,PATH_MAX);
  s->cwd = _tgetcwd(s->cwd,PATH_MAX);
  if (NULL == s->cwd)
    fatal_error(s,"%s: %s", __progname, strerror(errno));

  /* Anything left on the command line at this point is a file
     or directory we're supposed to process. If there's nothing
     specified, we should tackle standard input */
  
  if (optind == argc)
    hash_stdin(s);
  else
  {
    MD5DEEP_ALLOC(TCHAR,fn,PATH_MAX);

    count = optind;

    while (count < s->argc)
    {  
      generate_filename(s,fn,s->cwd,s->argv[count]);

#ifdef _WIN32
      status = process_win32(s,fn);
#else
      status = process_normal(s,fn);
#endif

     ++count;
    }

    free(fn);
  }
  
  if (primary_audit == s->primary_function)
    status = display_audit_results(s);

  return status;
}
